explorer/utils/dockerHubUtils.ts

```diff
@@ -105,29 +105,47 @@ export interface Manifest {
     schemaVersion: number;
 }
 
+export function isLoggedIn(): boolean {
+    return !!_token;
+}
+
 export async function dockerHubLogout(): Promise<void> {
     if (ext.keytar) {
         await ext.keytar.deletePassword(keytarConstants.serviceId, keytarConstants.dockerHubTokenKey);
         await ext.keytar.deletePassword(keytarConstants.serviceId, keytarConstants.dockerHubPasswordKey);
         await ext.keytar.deletePassword(keytarConstants.serviceId, keytarConstants.dockerHubUserNameKey);
     }
-    _token = null;
+
+    if (_token) {
+        ext.dockerExplorerProvider.refreshRegistries();
+        _token = null;
+    }
 }
 
 export async function dockerHubLogin(): Promise<{ username: string, password: string, token: string }> {
+    let id = await coreLogIn();
 
+    if (id && id.token && ext.keytar) {
+        await ext.keytar.setPassword(keytarConstants.serviceId, keytarConstants.dockerHubTokenKey, id.token);
+        await ext.keytar.setPassword(keytarConstants.serviceId, keytarConstants.dockerHubPasswordKey, id.password);
+        await ext.keytar.setPassword(keytarConstants.serviceId, keytarConstants.dockerHubUserNameKey, id.username);
+    }
+
+    return id;
+}
+
+async function coreLogIn(): Promise<{ username: string, password: string, token: string }> {
     const username: string = await vscode.window.showInputBox({ ignoreFocusOut: true, prompt: 'Please enter your Docker ID to log in to Docker Hub' });
     if (username) {
         const password: string = await vscode.window.showInputBox({ ignoreFocusOut: true, prompt: 'Please enter your Docker Hub password', password: true });
         if (password) {
             _token = await login(username, password);
             if (_token) {
+                ext.dockerExplorerProvider.refreshRegistries();
```

Comment:
It looks like you call `refreshRegistries` _before_ you set the keytar passwords in `dockerHubLogin`. I think you need to switch that order?

Reply:
Doesn't matter - key tar might not even be there.  It's _token that matters for the refresh.

Reply:
I don't see `_token` passed in to `refreshRegistries`, though. How is `_token` used in the refresh?

In `getDockerHubOrgs`, I only see us getting a token directly from keytar https://github.com/Microsoft/vscode-docker/blob/master/explorer/models/registryRootNode.ts#L73

Reply:
Okay, looks like you're right. At the moment it doesn't matter because the refresh is asynchronous. I have to rework this anyway as part of https://github.com/Microsoft/vscode-docker/issues/722.